src/memory_manager.rs

```diff
@@ -159,24 +182,46 @@ pub fn alloc<VM: VMBinding>(
     offset: usize,
     semantics: AllocationSemantics,
 ) -> Address {
-    // MMTk has assumptions about minimal object size.
-    // We need to make sure that all allocations comply with the min object size.
-    // Ideally, we check the allocation size, and if it is smaller, we transparently allocate the min
-    // object size (the VM does not need to know this). However, for the VM bindings we support at the moment,
-    // their object sizes are all larger than MMTk's min object size, so we simply put an assertion here.
-    // If you plan to use MMTk with a VM with its object size smaller than MMTk's min object size, you should
-    // meet the min object size in the fastpath.
-    debug_assert!(size >= MIN_OBJECT_SIZE);
-    // Assert alignment
-    debug_assert!(align >= VM::MIN_ALIGNMENT);
-    debug_assert!(align <= VM::MAX_ALIGNMENT);
-    // Assert offset
-    debug_assert!(VM::USE_ALLOCATION_OFFSET || offset == 0);
+    #[cfg(debug_assertions)]
+    crate::util::alloc::allocator::asset_allocation_args::<VM>(size, align, offset);
 
     mutator.alloc(size, align, offset, semantics)
 }
 
-/// Invoke the allocation slow path. This is only intended for use when a binding implements the fastpath on
+/// Allocate memory for an object.
+///
+/// The semantics of this function is the same as [`alloc`], except that when MMTk fails to allocate
+/// memory, it will simply return zero. This function is guaranteed not to trigger GC and not to
+/// call [`crate::vm::Collection::block_for_gc`] or [`crate::vm::Collection::out_of_memory`].
+///
+/// Generally [`alloc`] is preferred over this function. This function should only be used
+/// when the binding does not want GCs to happen at certain allocation sites (for example, places
+/// where stack maps cannot be generated), and is able to handle allocation failure if that happens.
+///
+/// Arguments:
+/// * `mutator`: The mutator to perform this allocation request.
+/// * `size`: The number of bytes required for the object.
+/// * `align`: Required alignment for the object.
+/// * `offset`: Offset associated with the alignment.
+/// * `semantics`: The allocation semantic required for the allocation.
+pub fn alloc_no_gc<VM: VMBinding>(
```

Comment:
What are the semantics of `alloc_no_gc` with `oom_on_acquire`, i.e. what happens when you try to allocate a massive (let's say 25 TB) object? Is it possible to add this as a test?

Reply:
Good point. I made some changes so we won't call `out_of_memory` when the `will_oom_on_acquire` check fails. I also added a test for it.